### PR TITLE
Add tokenizer's chat template to SFT

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description
* This PR standardizes how we format conversational data for SFT. Our previous custom chat template was causing issues during fine-tuning of student model for distillation because the distillation data generation script uses the tokenizer's specific chat template. To ensure consistency and avoid unexpected results, we're modifying the SFT code to also adopt the tokenizer's chat template.
* This PR also reduces the golden data file size from 75M to 18M.

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/configs/sft.yml [hunk 1]
```diff
@@ -14,6 +14,9 @@
 
 base_config: "base.yml"
 
+enable_goodput_recording: False
+monitor_goodput: False
```
Comment (bvandermoon) [inline MaxText/configs/sft.yml:18]: This will be the default in `base.yml`: https://github.com/AI-Hypercomputer/maxtext/pull/1750. Probably good to just remove here

**Review by bvandermoon — COMMENTED** [review]
> we're modifying the SFT code to also adopt the tokenizer's chat template

Could you add the template to the PR description? It would help for context while reviewing

**Comment by SurbhiJainUSC** [comment]
> > we're modifying the SFT code to also adopt the tokenizer's chat template
> 
> Could you add the template to the PR description? It would help for context while reviewing

The template is different for each model. I will add an example maybe for context.

File: MaxText/configs/sft.yml (hunk 1 above)
Comment (SurbhiJainUSC) [reply MaxText/configs/sft.yml]: Done!

File: end_to_end/tpu/test_sft_trainer.sh [hunk 2]
```diff
@@ -45,16 +45,13 @@ sorted_dirs=($(printf '%s\n' "${integer_dirs[@]}" | sort -n))
...
-PROMPT="<user>${PROMPT}</user> <assistant>"
-
 # Decode
 python3 -m MaxText.decode MaxText/configs/sft.yml \
     run_name=${RUN_NAME}-hf-decode \
     model_name=${PRE_TRAINED_MODEL} tokenizer_path=${PRE_TRAINED_MODEL_TOKENIZER} tokenizer_type=huggingface \
     load_parameters_path=${FINE_TUNED_MODEL_CKPT_PATH} \
     per_device_batch_size=${PER_DEVICE_BATCH_SIZE} max_target_length=128 max_prefill_predict_length=64 \
-    attention=dot_product decode_sampling_strategy=greedy prompt="${PROMPT}"
+    attention=dot_product decode_sampling_strategy=greedy prompt="${PROMPT}" use_chat_template=True
```
Comment (gagika) [inline end_to_end/tpu/test_sft_trainer.sh:54]: besides chat template, does SFT support other instruction tuning (or custom) templates?

File: MaxText/input_pipeline/_hf_data_processing.py [hunk 3]
```diff
@@ -64,6 +64,14 @@ def preprocessing_pipeline(
   if shuffle:
     dataset = dataset.shuffle(seed=data_shuffle_seed)
 
+  tokenizer = transformers.AutoTokenizer.from_pretrained(
+      tokenizer_path,
+      add_bos_token=add_bos if not use_sft else False,
+      add_eos_token=add_eos if not use_sft else False,
```
Comment (gagika) [inline MaxText/input_pipeline/_hf_data_processing.py:70]: will `bos_token` and `eos_token` be added later for SFT? what's the reason of having False here?

File: MaxText/scratch_code/generate_sft_in_trl_golden_logits.py [hunk 4]
```diff
@@ -71,41 +71,26 @@ def setup_sft_trainer(data, hf_model, tokenizer, max_target_length):
...
-  input_ids = [tokenizer.bos_token_id] + input_ids + [tokenizer.eos_token_id]
-  attention_mask = [1] + attention_mask + [1]
-  return {
-      "input_ids": torch.tensor(input_ids[:max_target_length], dtype=torch.long).unsqueeze(0),
-      "labels": torch.tensor(labels[:max_target_length], dtype=torch.long).unsqueeze(0),
-      "attention_mask": torch.tensor(attention_mask[:max_target_length], dtype=torch.long).unsqueeze(0),
-  }
+def prepare_trl_inputs(tokenizer, max_target_length):
+  """Get tokenized inputs."""
+  data_in_chat_format = tokenizer.apply_chat_template(DATA["messages"], tokenize=False)
```
Comment (hengtaoguo) [inline MaxText/scratch_code/generate_sft_in_trl_golden_logits.py:76]: Just for my understanding. For some tokenizers from HuggingFace, they have this `apply_chat_template` as tokenizer's built-in function? And different models' tokenizer may return different template?

**Comment by hengtaoguo** [comment]
> > > we're modifying the SFT code to also adopt the tokenizer's chat template
> > 
> > 
> > Could you add the template to the PR description? It would help for context while reviewing
> 
> The template is different for each model. I will add an example maybe for context.

+1 thank you and that would be very helpful. Something like (1) a raw data example from dataset; (2) transformed prompt after using a specific tokenizer's apply_chat_template.

**Review by hengtaoguo — APPROVED** [review]

File: end_to_end/tpu/test_sft_trainer.sh (hunk 2 above)
Comment (SurbhiJainUSC) [reply end_to_end/tpu/test_sft_trainer.sh:54]: That's not currently supported, but we can definitely add that feature if it would be helpful to users.

File: MaxText/input_pipeline/_hf_data_processing.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply MaxText/input_pipeline/_hf_data_processing.py:70]: Chat template already a bos and a eos token. 
Chat template applied to `messages = [{"role": "user", "content": "Hi, how are you?"}, {"role": "assistant", "content": "Hi, I am good."}]` using "deepseek-ai/DeepSeek-V2-Lite-Chat" tokenizer:

```
<｜begin▁of▁sentence｜>User: Hi, how are you?

Assistant: Hi, I am good.<｜end▁of▁sentence｜>
```

File: MaxText/input_pipeline/_input_pipeline_utils.py [hunk 5]
```diff
@@ -101,18 +101,48 @@ def is_conversational(features, data_columns):
...
+        prompt_completion_tokens = tokenizer.apply_chat_template(
+            [prompt, message], add_generation_prompt=False, tokenize=True
+        )
+        prompt_tokens = tokenizer.apply_chat_template([prompt], add_generation_prompt=False, tokenize=True)
+        completion_tokens = prompt_completion_tokens[len(prompt_tokens) :]
+        completion_in_chat_template = tokenizer.decode(completion_tokens, skip_special_tokens=False)
+        messages.append(completion_in_chat_template)
+        is_prompt.append(False)
+  except ValueError as e:
+    max_logging.log(f"Unable to apply chat template: {e}")
```
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply MaxText/input_pipeline/_input_pipeline_utils.py:144]: When we are fine-tuning with SFT, we should only use instruction-tuned/chat model and tokenizer for that model has a chat template specified already. If not, then we should just exit unless we add a support in future for users to provide a custom template on their end.

File: MaxText/input_pipeline/_input_pipeline_utils.py [hunk 6]
```diff
@@ -101,18 +101,48 @@ def is_conversational(features, data_columns):
...
-      is_prompt.append(True)
-    elif x["role"] == "assistant":
-      messages.append("<assistant>" + x["content"] + "</assistant>")
-      is_prompt.append(False)
+  prompt = None
+  try:
+    for message in example[data_column_name]:
+      if message["role"] == "user":
+        prompt = message
+        prompt_in_chat_template = tokenizer.apply_chat_template([prompt], add_generation_prompt=False, tokenize=False)
```
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply MaxText/input_pipeline/_input_pipeline_utils.py:131]: Yes, each model can have a different chat template.

**Review by bvandermoon — APPROVED** [review]
